spec/unit/pops/types/p_object_type_spec.rb

```diff
@@ -1136,6 +1140,282 @@ def compile(code)
+        notice($x =~ MyType)
+        notice($x =~ MyType[p1 => 'hello', p2 => 'cruel world'])
+        notice($x =~ MyType[p1 => 'good bye', p2 => 'cruel world'])
+        PUPPET
+      end
+
+      it 'no constraint != default constrant' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Enum['hello', 'good bye'] },
+            p2 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Pattern[/world/, /universe/] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('greetings', 'earthlings')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        PUPPET
+      end
+
+      it 'Undef can be used to prevent attribute setting' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'true', 'true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]], value => Enum['hello', 'good bye'] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]], value => Pattern[/world/, /universe/] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        notice($x =~ MyType[default, default])
+        notice($x =~ MyType[default, undef])
+        PUPPET
+      end
+
+      it 'Required parameters cannot be set using default' do
+        expect{eval_and_collect_notices(<<-PUPPET, node)}.to raise_error(/No value provided for required type_parameter MyType\[p1\]/)
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default, default])
+        PUPPET
+      end
+
+      it 'Required parameters must be provided' do
+        expect{eval_and_collect_notices(<<-PUPPET, node)}.to raise_error(/No value provided for required type_parameter MyType\[p2\]/)
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
```

Comment:
this line should probably not be here as it is supposed to succeed, and if there is a "no value provided error" for this line it is taken as success even though this is not what is intended. (Same as comment on test above).